examples/Upscaling0FormMARS.cpp

```diff
@@ -0,0 +1,604 @@
+
+int main (int argc, char *argv[])
+{
+    // 1. Initialize MPI
+    mpi_session sess(argc,argv);
+    MPI_Comm comm = MPI_COMM_WORLD;
+
+    int num_procs, myid;
+    MPI_Comm_size(comm, &num_procs);
+    MPI_Comm_rank(comm, &myid);
+
+    const char* meshfile_c = "../meshes/boxcyl.mesh3d";
+    int ser_ref_levels = 0;
+    int par_ref_levels = 1;
+    int feorder = 0;
+    int upscalingOrder = 0;
+    bool do_visualize = true;
+    OptionsParser args(argc, argv);
+    args.AddOption(&meshfile_c, "-m", "--mesh",
+                   "MFEM mesh file to load.");
+    args.AddOption(&ser_ref_levels, "-sr", "--nref_serial",
+                   "Number of times to refine serial mesh.");
+    args.AddOption(&par_ref_levels, "-pr", "--nref_parallel",
+                   "Number of times to refine parallel mesh.");
+    args.AddOption(&feorder, "-feo", "--feorder",
+                   "Polynomial order of fine finite element space.");
+    args.AddOption(&upscalingOrder, "-uo", "--upscalingorder",
+                   "Target polynomial order of coarse space.");
+    args.AddOption(&do_visualize, "-v", "--do-visualize", "-nv", "--no-visualize",
+                   "Do interactive GLVis visualization.");
+    args.Parse();
+    if (!args.Good())
+    {
+        args.PrintUsage(std::cout);
+        MPI_Finalize();
+        return 1;
+    }
+    PARELAG_ASSERT(args.Good());
+    std::string meshfile(meshfile_c);
+
+    // default linear solver options
+    constexpr int print_iter = 0;
+    constexpr int max_num_iter = 500;
+    constexpr double rtol = 1e-6;
+    constexpr double atol = 1e-12;
+
+    if (myid == 0)
+    {
+        std::cout << "Read mesh " << meshfile << std::endl;
+        std::cout << "Finite Element Order " << feorder << std::endl;
+        std::cout << "Upscaling Order " << upscalingOrder << std::endl;
+        std::cout << "Refine mesh in serial "
+                  << ser_ref_levels << " times.\n";
+        std::cout << "Refine mesh in parallel "
+                  << par_ref_levels << " times.\n";
+    }
+
+    shared_ptr<Mesh> mesh;
+    shared_ptr<Mesh> tmesh;
+    Array<int> ess_attr;
+    {
+        std::ifstream imesh(meshfile.c_str());
+        if (!imesh)
+        {
+            if (myid == 0)
+                std::cerr << std::endl << "Can not open mesh file: "
+                          << meshfile << std::endl << std::endl;
+            return EXIT_FAILURE;
+        }
+
+        mesh = make_shared<Mesh>(imesh, 1, 1);
+        imesh.close();
+
+        ess_attr.SetSize(mesh->bdr_attributes.Max());
+        ess_attr = 1;
+
+        for (int l = 0; l < ser_ref_levels; l++)
+            mesh->UniformRefinement();
+    }
+
+    const int nDimensions = mesh->Dimension();
+    const int geom = mesh->GetElement(0)->GetType();
+    const int bdr_geom = (mesh->GetNBE() == 0)? 0 : mesh->GetBdrElement(0)->GetType();
+    const int nLevels = par_ref_levels+1;
+    Array<int> *partitions[nLevels - 1];
+    mars::Mesh2 mesh2;
```

Comment:
That would be the proper thing to do. Currently, the fate of this is not fully clear and is a WIP. Particularly, MARS is not a de facto parallel library and this doesn't align well with the rest or ParElag. So I would like to keep everything MARS related a bit separate until it gets more established. In this sense, it is useful but a bit artificial at this point. Thus, it would be best to not yet interfere with the preexisting functionality and code in ParElag.

Reply:
Okay, that sounds fine.